Assets/Scripts/Layers/Adapters/DataTypeAdapters/WFSImportAdapter.cs

```diff
@@ -0,0 +1,292 @@
+        {
+            // Find the <ows:Operation name="GetFeature"> node
+            var getFeatureOperationNode = xmlDocument.SelectSingleNode("//ows:Operation[@name='GetFeature']", namespaceManager);
+            
+            if (getFeatureOperationNode == null)
+                Debug.LogWarning("WFS GetFeature operation not found.");
+
+            return getFeatureOperationNode;
+        }
+
+        private string[] GetFeatureTypes(LocalFile localFile)
+        {
+            // Read the XML data to find the list of feature types
+            var cachedDataPath = localFile.LocalFilePath;
+            var dataAsText = File.ReadAllText(cachedDataPath);
+            var xmlDocument = new XmlDocument();
+            xmlDocument.LoadXml(dataAsText);
+
+            XmlNamespaceManager namespaceManager = ReadNameSpaceManager(xmlDocument);
+
+            // Find the FeatureTypeList node somewhere in xmldocument(that might not start with wfs:)
+            var featureTypeListNodeInRoot = xmlDocument.SelectSingleNode("//*[local-name()='FeatureTypeList']", namespaceManager);
+            var featureTypeChildNodes = featureTypeListNodeInRoot.ChildNodes;
+            var featureTypes = new List<string>();
+
+            foreach(XmlNode featureTypeNode in featureTypeChildNodes)
+            {
+                var featureTypeName = featureTypeNode.SelectSingleNode(".//*[local-name()='Name']", namespaceManager).InnerText;
+                featureTypes.Add(featureTypeName);
+            }
+
+            return featureTypes.ToArray();
+        }
+
+        private void AddWFSLayer(string featureType, string sourceUrl, FolderLayer folderLayer)
+        {
+            Debug.Log("Adding WFS layer: " + featureType);
+
+            // Start by removing any query parameters we want to inject
+            var uriBuilder = new UriBuilder(sourceUrl);
+
+            // Make sure we have a wfsVersion set
+            if(string.IsNullOrEmpty(wfsVersion))
+            {
+                Debug.LogWarning("WFS version could not be determined, defaulting to " + defaultFallbackVersion);
+                wfsVersion = defaultFallbackVersion;
+            }
+
+            // Set the required query parameters for the GetFeature request
+            uriBuilder.SetQueryParameter("service", "WFS");
+            uriBuilder.SetQueryParameter("request", "GetFeature");
+            uriBuilder.SetQueryParameter("version", wfsVersion);
+            uriBuilder.SetQueryParameter("typeNames", featureType);
+            uriBuilder.SetQueryParameter("outputFormat", geoJsonOutputFormat);
+            uriBuilder.SetQueryParameter("bbox", "{0}"); // Bbox value is injected by CartesianTileWFSLayer
+
+            var getFeatureUrl = uriBuilder.Uri.ToString();
+
+            // Create a new GeoJSON layer per GetFeature, with a 'live' datasource
+            var layerGameObject = new GameObject(featureType);
+            var layer = layerGameObject.AddComponent<GeoJSONLayer>();
```

Comment:
Isn't this more or less a duplicate of the GeoJSONAdapter? The cool thing with a Chain of Responsibility pattern with a Strategy pattern is that you can call the Chain again and make use of composition. In this case, you could save the contents of the GeoJSON to disk, make a LocalFile and call the chain again with that file. That would allow for future GML support by just adding a new adapter

Reply:
It also removes duplicated things, and makes changes to the GeoJSON import process easier to adapt (see what I did there? :D )

Reply:
Well in this case there isnt realy a 'file'. The WFSGeoJSONTileDataLayer component I add below this step provides the data for that layer, so the composition happens here, a composition very specific to a WFS.

The only duplication I see here is that both adapters create a GeoJSONLayer that happens to require the SetDefaultVisualizerSettings method to be called for it to work. Removing that requirement there would remove the need for the duplication here. (having to plug in the same prefabs for the visualisers)

Reply:
Instead of new GameObjects() layers I'll use prefabs. Folder structure something like LayerPrefabs. This removes specific configs from the adapters, and can be placed on the prefabs ( like random colors )